### PR TITLE
fix: v1 templates with metadata + always cleanup orphaned secrets

### PR DESCRIPTION
**Pull request by thesuperzapper**
## Problem Statement

This PR makes the following changes:

1. Fixes an issue that prevented v1 templates from defining metadata/annotations (I think this was present before 0.11):
     - https://github.com/external-secrets/external-secrets/issues/4172
     - https://github.com/external-secrets/external-secrets/issues/1910
     - __NOTE:__ this PR also updates the docs for v1 templating to explain that annotations/labels are not templated in v1
3. We now always check if there are orphaned secrets when `creationPolicy=Owner`:
      - Previously, we only checked when we first created a new target secret, but if that failed, we would not retry on the next reconcile.
      - Because we are using the partial secret cache to check for orphaned secrets, it should not have a significant performance penalty, and will not make any new API call.
4. Improves the secret key validation in `dataFrom`, to also prevent keys being longer than 253 characters.
5. Improves the error messages when there is an error in one of the `spec.dataFrom[]` elements 
6. We now correctly set the `Reason` of Events for ExternalSecrets in following situations:
     - `Created` New target Secret created
     - `Deleted` Existing target Secret deleted
     - `Updated` Existing target Secret updated
     - `MissingProviderSecret` when `deletionPolicy != Retain` and an expected provider secret does not exist
         - Previously this was incorrectly marked as `Deleted`, even though the target secret was not being deleted.

## Related Issue

fixes https://github.com/external-secrets/external-secrets/issues/4172

## Checklist

- [X] I have read the [contribution guidelines](https://external-secrets.io/latest/contributing/process/#submitting-a-pull-request)
- [X] All commits are signed with `git commit --signoff`
- [X] My changes have reasonable test coverage
- [X] All tests pass with `make test`
- [X] I ensured my PR is ready for review with `make reviewable`

### CODE REVIEW

File: pkg/controllers/externalsecret/externalsecret_controller.go [hunk 1]
```diff
@@ -333,17 +341,19 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (result ct
...
+				err = fmt.Errorf(errDeleteCreatePolicy, secretName, creationPolicy)
 				r.markAsFailed(msgErrorDeleteSecret, err, externalSecret, syncCallsError.With(resourceLabels))
 				return ctrl.Result{}, nil
 			}
 
 			// delete the secret, if it exists
 			if existingSecret.UID != "" {
-				if err := r.Delete(ctx, existingSecret); err != nil && !apierrors.IsNotFound(err) {
+				err = r.Delete(ctx, existingSecret)
+				if err != nil && !apierrors.IsNotFound(err) {
```
Comment (Skarlso) [inline pkg/controllers/externalsecret/externalsecret_controller.go:352]: Why did you remove the inlining? 🤔

Comment (thesuperzapper) [reply pkg/controllers/externalsecret/externalsecret_controller.go:352]: It felt easier to read

Comment (thesuperzapper) [reply pkg/controllers/externalsecret/externalsecret_controller.go:352]: Especially with the longer mutli `&&`

Comment (Skarlso) [reply pkg/controllers/externalsecret/externalsecret_controller.go:352]: okay

File: pkg/controllers/externalsecret/externalsecret_controller_secret.go [hunk 2]
```diff
@@ -110,83 +123,108 @@ func toStoreGenSourceRef(ref *esv1beta1.StoreSourceRef) *esv1beta1.StoreGenerato
...
 	}
 	if len(remoteRef.Rewrite) == 0 {
-		// ConversionStrategy is deprecated. Use RewriteMap instead.
-		r.recorder.Event(externalSecret, v1.EventTypeWarning, esv1beta1.ReasonDeprecated, fmt.Sprintf("dataFrom[%d].find.conversionStrategy=%v is deprecated and will be removed in further releases. Use dataFrom.rewrite instead", i, remoteRef.Find.ConversionStrategy))
 		secretMap, err = utils.ConvertKeys(remoteRef.Find.ConversionStrategy, secretMap)
 		if err != nil {
-			return nil, fmt.Errorf(errConvert, err)
+			return nil, fmt.Errorf(errConvert, remoteRef.Find.ConversionStrategy, err)
 		}
 	}
```
Comment (gusfcarvalho) [inline pkg/controllers/externalsecret/externalsecret_controller_secret.go:216]: We can probably just remove this check and remove ConversionStrategy as part of this PR.

It has been deprecated for a while now.

File: pkg/controllers/externalsecret/externalsecret_controller_secret.go [hunk 3]
```diff
@@ -110,83 +123,108 @@ func toStoreGenSourceRef(ref *esv1beta1.StoreSourceRef) *esv1beta1.StoreGenerato
...
-		return nil, fmt.Errorf(errRewrite, i, err)
+		return nil, fmt.Errorf(errRewrite, err)
 	}
 	if len(remoteRef.Rewrite) == 0 {
 		secretMap, err = utils.ConvertKeys(remoteRef.Extract.ConversionStrategy, secretMap)
 		if err != nil {
-			return nil, fmt.Errorf(errConvert, err)
+			return nil, fmt.Errorf(errConvert, remoteRef.Extract.ConversionStrategy, err)
 		}
 	}
```
Comment (gusfcarvalho) [inline pkg/controllers/externalsecret/externalsecret_controller_secret.go:176]: Same here - we might just remove ConversionStrategy

File: pkg/controllers/externalsecret/externalsecret_controller_secret.go (hunk 2 above)
Comment (thesuperzapper) [reply pkg/controllers/externalsecret/externalsecret_controller_secret.go:216]: @gusfcarvalho I don't think we can remove it, because it's used by default on all providers that dont use rewrite, and is important to prevent creating invalid keys, there is no clear replacement default behavior.

All I am doing by removing this event creation here is preventing random Event API spam for every loop.

If we want to actually warn users about its deprecation, we should do it with admission webhook warnings, and also have a replacement, because all we are doing by "removing it" is saying that the user has no say, and we will use the "default" conversion strategy for invalid keys.

My preference is that we dont deprecate this.

File: pkg/controllers/externalsecret/externalsecret_controller_secret.go (hunk 3 above)
Comment (thesuperzapper) [reply pkg/controllers/externalsecret/externalsecret_controller_secret.go:176]: See https://github.com/external-secrets/external-secrets/pull/4174#discussion_r1873983817

**Comment by thesuperzapper** [comment]
@gusfcarvalho @Skarlso I am wanting to raise another PR which is based on this one (to avoid merge conflicts).

As these changes are not breaking, and only fix bugs or improve logs/messages, can we take another look at merging this soon?

File: pkg/controllers/externalsecret/externalsecret_controller_secret.go (hunk 2 above)
Comment (thesuperzapper) [reply pkg/controllers/externalsecret/externalsecret_controller_secret.go:216]: The main reason I am making a change to this line of code is to stop creating an Event on every dataFrom.Find that does not set rewrite.

File: pkg/controllers/externalsecret/externalsecret_controller.go [hunk 4]
```diff
@@ -592,16 +610,18 @@ func deleteOrphanedSecrets(ctx context.Context, cl client.Client, externalSecret
...
-	if err := cl.List(ctx, secretListPartial, listOpts); err != nil {
+	if err := r.List(ctx, secretListPartial, listOpts); err != nil {
 		return err
 	}
 
 	// delete all secrets that are not the target secret
 	for _, secretPartial := range secretListPartial.Items {
 		if secretPartial.GetName() != secretName {
-			if err := cl.Delete(ctx, &secretPartial); err != nil {
+			err := r.Delete(ctx, &secretPartial)
```
Comment (Skarlso) [inline pkg/controllers/externalsecret/externalsecret_controller.go:620]: This one was reading better with inlining though. :P

File: pkg/template/engine.go [hunk 5]
```diff
@@ -30,8 +32,5 @@ func EngineForVersion(version esapi.TemplateEngineVersion) (ExecFunc, error) {
 	case esapi.TemplateEngineV2:
 		return v2.Execute, nil
 	}
-
-	// in case we run with a old v1alpha1 CRD
-	// we must return v1 as default
-	return v1.Execute, nil
+	return nil, fmt.Errorf("unsupported template engine version: %s", version)
```
Comment (thesuperzapper) [inline pkg/template/engine.go:35]: This change might not be correct @Skarlso, because I guess its possible that if someone upgrades from `v1alpha1` to `v1beta1` it might not correctly set the default.

I might need to rever this part.

**Review by Skarlso — APPROVED** [review]

File: pkg/template/engine.go (hunk 5 above)
Comment (thesuperzapper) [reply pkg/template/engine.go:35]: I made a PR to fix this, while still keeping the check for actually invalid values: https://github.com/external-secrets/external-secrets/pull/4182
